### PR TITLE
Add a progress bar for project loading, saving, etc.

### LINKED ISSUE #152: [New addon] Progress bar for project loading and saving
**Describe the new addon**

Sometimes projects can take a long time to load and save, so having a progress bar would be useful.

### PR DESCRIPTION
**Pull request by GarboMuffin**
**Resolves**

Resolves #152

**Changes**

This adds a progress bar for project loading, saving, etc.

The progress bar is on the top of the page. It's a white bar. It looks like this:

image

A full list of the things that are monitored:

 - loading project.json (based on download size)
 - loading project assets (based on number of assets loaded and not yet loaded, not actual download size, just like Scratch 2)
 - uploading project (based on upload size)
 - uploading assets (based on number of assets uploaded and not yet uploaded, not actual upload size.)
 - remixing, copying (based on upload size)

The first 2 each being monitored means that you might see 2 progress bars complete during page loading, one after the other. That's probably not optimal.

Not monitored:

 - uploading project thumbnail (happens after project is uploaded, usually very small and generally doesn't get larger over time)

**Reason for changes**

I have a project that takes like 40 seconds to save and load, so having a progress bar to get an idea as to how long it will take would be useful. Decided to add progress bars for everything else too.

**Tests**

This hasn't been thoroughly tested, but it works for me on all the projects I've tried.

This uses a lot of terrible prototype hacks to function, although I've tried to make sure they're safe and don't break anything.

### CODE REVIEW

**Comment by apple502j** [comment]
I am re-implementing this, with less prototype hacks.

**Comment by GrahamSH-LLK** [comment]
> ~I am re-implementing this, with less prototype hacks.~ seems like impossible

That's not good, why?

**Comment by WorldLanguages** [comment]
I'll take a look at this later, but let me mention I dislike the idea of having addons replace browser APIs, like `fetch` and `XMLHttpRequest`. Too many things can go wrong-

**Review by WorldLanguages — APPROVED** [review]
I dislike prototype hacks (specially if addons make them, not the addon loader) but it's fine.

**Comment by tb148** [comment]
Should we merge it? It have an approving review (by @WorldLanguages) and prettier checks also passed.

**Comment by Hans5958** [comment]
Do it later? Wanted to talk about thinning it.

**Comment by apple502j** [comment]
Do we really need to check for the progress? Isn't number of asset loaded enough?

**Comment by Hans5958** [comment]
You see, there is a (size) difference between a simple `.svg` image and a huge audio file. So...

**Comment by NoobTracker** [comment]
imo that bar should be where the normal loading screen is, that would look better.

**Comment by Hans5958** [comment]
I was thought the same, but, you see, it's not just for project loading, and it is made so it works inside the editor.

**Comment by NoobTracker** [comment]
@Hans5958 Well, in the editor you also have that big, blue loading screen, right?
And for saving there could just be a progress bar where you normally only have that loading symbol.

**Comment by NoobTracker** [comment]
> 
> 
> Not when you saving a project, and some other actions.



> 
> 
> And for saving there could just be a progress bar where you normally only have that loading symbol.

**Comment by Hans5958** [comment]
I think it is too small, and I think it's for consistency sake.

**Comment by NoobTracker** [comment]
I want that the progress bar is exactly there where something is actually doing something, not at the top.
And you could make it bigger.

**Comment by GarboMuffin** [comment]
I put the progress bar on the top of the page because it was the easiest thing to do.

I'll look into putting the progress bar in the relevant spots later today, don't merge yet.

**Comment by GarboMuffin** [comment]
Uq71jFcWNO

This is what the loading progress bar looks like now. I think I'll replace Scratch's "Creating blocks" "Loading sprites" "Herding cats" text with "X/Y assets loaded"

Saving, remixing, copying still uses a progress bar on the top of the screen (and are probably very broken)

**Comment by apple502j** [comment]
One edit conflict.

**Comment by GarboMuffin** [comment]
Loading project

image

Saving, copying

image


Remixing doesn't work yet.

**Comment by Hans5958** [comment]
@GarboMuffin 

How about making it so there are two options, the top, [PACE](https://github.hubspot.com/pace/docs/welcome/)-styled one, and the integrated part, the current design?

Also, this question, for the top bar progress, just in case it is implemented.

> Just a suggestion, how about making it thinner, like 2px?

**Comment by GarboMuffin** [comment]
@Hans5958 I'll add an option to make the progress bar always be at the top of the page, like it was previously.

Remixing from the project page

image

Remixing from the editor

image

**Comment by NoobTracker** [comment]
That looks really good so far.

**Comment by GarboMuffin** [comment]
This is ready to be reviewed and merged if there are no significant bugs.

**Comment by WorldLanguages** [comment]
First `positive_integer` setting! Nice

**Review by Hans5958 — APPROVED** [review]
Works on Firefox 80.0.1 (64-bit)

File: addons/progress-bar/userscript.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    originalPostMessage.call(this, message, options);
+  };
+
+  function inject() {
+    // When the progress bar is already in the document, we do not need to do anything.
+    if (document.querySelector(".u-progress-bar-outer")) {
+      return;
+    }
+
+    const loaderMessageContainerOuter = document.querySelector("[class^=loader_message-container-outer]");
```
Comment (apple502j) [inline addons/progress-bar/userscript.js:200]: selectors need ""

File: addons/progress-bar/userscript.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    if (loaderMessageContainerOuter) {
+      loaderMessageContainerOuter.hidden = true;
+      loaderMessageContainerOuter.parentElement.appendChild(loadingCaption);
+      loaderMessageContainerOuter.parentElement.appendChild(barOuter);
+      return;
+    }
+
+    const spinner = document.querySelector("[class^=inline-message_spinner]");
```
Comment (apple502j) [inline addons/progress-bar/userscript.js:208]: here too add ""

File: addons/progress-bar/userscript.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+      return;
+    }
+
+    const remixButton = document.querySelector(".remix-button.remixing");
+    if (remixButton) {
+      remixButton.appendChild(barOuter);
+      return;
+    }
+
+    const alertMessage = document.querySelector("[class^=alert_alert-message] span");
```
Comment (apple502j) [inline addons/progress-bar/userscript.js:221]: here again, attribute selector needs ""

**Comment by WorldLanguages** [comment]
@GarboMuffin, please tell me if you've finished making changes. If so, I'll resolve the conflicts and merge.

**Comment by GarboMuffin** [comment]
I'm done.

**Pull request merged**